[homematic] Support sending datapoints via different rx modes

FStolte opened the pull request:
Provides an endpoint in the HomematicThingHandler for specifying the rx mode when sending datapoints to the homematic gateway. This allows custom ThingHandlers in ESH-based solutions to change the used rxMode from the default one.

Th rx mode can be configured as "BURST" or "WAKEUP". In BURST mode, the datapoint is transmitted to the device immediately by waking it up through burst messages. This, however, requires more duty cycle than a normal message. In WAKEUP mode, the message is only sent once the device wakes up by itself and contacts the homematic gateway.

Signed-off-by: Florian Stolte <fstolte@itemis.de>

htreu commented on extensions/binding/org.eclipse.smarthome.binding.homematic/src/main/java/org/eclipse/smarthome/binding/homematic/handler/HomematicThingHandler.java:
```diff
@@ -300,6 +301,22 @@ public void handleCommand(ChannelUID channelUID, Command command) {
+     * @return The rxMode ({@link HomematicBindingConstants#RX_BURST_MODE "BURST"} for burst mode,
+     *         {@link HomematicBindingConstants#RX_WAKEUP_MODE "WAKEUP"} for wakeup mode, or null for the default mode)
+     */
+    protected String getRxMode() {
```
This implies a handler will always send all commands either in BURST or WAKEUP mode. Is this reality or a simplification?

FStolte replied:
Generally, the method might return different rx modes under different conditions. In our solution, e.g., we plan to switch to WAKEUP mode for certain device types when we already have a high duty cycle on the bridge. Thn, we switch back to BURST mode when the duty cycle drops.

htreu replied:
So with this it's not possible to specifically define the rx mode for a single call to `sendDatapoint`. Is this a problem? Maybe even during high duty cycle you want to issue a specific command or request in BURST mode but due to multi threading you can not be sure your call will be executed in BURST mode. Or is this only academic?

FStolte replied:
I think I see where you're coming from. The solution would be much more flexible if the logic for selecting the rx mode would have access to everything that's passed to `sendDatapoint()`. The reason why I didn't just make `sendDatapoint()` protected (which would provide this flexibility) is that it can throw the bundle-internal `HomematicClientException`.

Passing the datapoint, config and newValue parameters from the `sendDatapoint()` call to the `getRxMode()` call should provide sufficiently flexibility, do you agree? (For our solution, the current approach would be enough, but I see how at least the datapoint might be interesting for the decision of the rx mode)

Another solution would be to expose the `HomematicClientException` through an exported package and then making `sendDatapoint()` protected, but I tried to avoid that.

FStolte commented:
I have added a commit with an improved version of the getRxMode() method to provide more flexibility for the implementor of that method.

htreu commented:
Unfortunately javac disagrees:

```
[ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:1.2.0:compile (default-compile) on project org.eclipse.smarthome.binding.homematic: Compilation failure: Compilation failure: 
[ERROR] /home/travis/build/eclipse/smarthome/extensions/binding/org.eclipse.smarthome.binding.homematic/src/main/java/org/eclipse/smarthome/binding/homematic/internal/communicator/virtual/OnTimeAutomaticVirtualDatapointHandler.java:[81] 
[ERROR] 	gateway.sendDatapoint(dpOnTime, new HmDatapointConfig(), getVirtualDatapointValue(channel));
[ERROR] 	        ^^^^^^^^^^^^^
[ERROR] The method sendDatapoint(HmDatapoint, HmDatapointConfig, Object, String) in the type HomematicGateway is not applicable for the arguments (HmDatapoint, HmDatapointConfig, Double)
[ERROR] /home/travis/build/eclipse/smarthome/extensions/binding/org.eclipse.smarthome.binding.homematic/src/main/java/org/eclipse/smarthome/binding/homematic/internal/communicator/virtual/DisplayTextVirtualDatapoint.java:[376] 
[ERROR] 	gateway.sendDatapoint(channel.getDatapoint(HmParamsetType.VALUES, DATAPOINT_NAME_SUBMIT),
[ERROR] 	        ^^^^^^^^^^^^^
[ERROR] The method sendDatapoint(HmDatapoint, HmDatapointConfig, Object, String) in the type HomematicGateway is not applicable for the arguments (HmDatapoint, HmDatapointConfig, String)
[ERROR] /home/travis/build/eclipse/smarthome/extensions/binding/org.eclipse.smarthome.binding.homematic/src/main/java/org/eclipse/smarthome/binding/homematic/internal/communicator/virtual/DisplayOptionsVirtualDatapointHandler.java:[83] 
[ERROR] 	gateway.sendDatapoint(dp, new HmDatapointConfig(), newValue);
[ERROR] 	        ^^^^^^^^^^^^^
[ERROR] The method sendDatapoint(HmDatapoint, HmDatapointConfig, Object, String) in the type HomematicGateway is not applicable for the arguments (HmDatapoint, HmDatapointConfig, Object)
```

FStolte commented:
I had adjusted the calls for the virtual datapoints previously. However, the changes must've gotten lost when I reworked my initial implementation idea...

Anyway, now the PR is complete. The additional calls from the virtual datapoints simply use the default rx mode, so their behavior isn't changed.
